internal/topology/check/compatibility.go

```diff
@@ -18,39 +18,179 @@ limitations under the License.
+			field.NewPath("metadata", "name"),
+			desired.GetName(),
+			fmt.Sprintf("must be equal to %v for %v", current.GetName(), desired.GetName()),
+		))
 	}
-	return ReferencedObjectsAreCompatible(current, desired)
+	allErrs = append(allErrs, ReferencedObjectsAreCompatible(current, desired)...)
+	return allErrs
 }
 
 // ReferencedObjectsAreCompatible checks if two referenced objects are compatible, meaning that
 // they are of the same GroupKind and in the same namespace.
-func ReferencedObjectsAreCompatible(current, desired client.Object) error {
+func ReferencedObjectsAreCompatible(current, desired client.Object) field.ErrorList {
+	var allErrs field.ErrorList
+
 	currentGK := current.GetObjectKind().GroupVersionKind().GroupKind()
 	desiredGK := desired.GetObjectKind().GroupVersionKind().GroupKind()
-
-	if currentGK.String() != desiredGK.String() {
-		return errors.Errorf("invalid operation: it is not possible to change the GroupKind of %s/%s from %s to %s",
-			current.GetObjectKind().GroupVersionKind(), current.GetName(), currentGK, desiredGK)
+	if currentGK.Group != desiredGK.Group {
+		allErrs = append(allErrs, field.Invalid(
+			field.NewPath("metadata", "apiVersion"),
+			desiredGK.Group,
+			fmt.Sprintf("api group must be equal to %s for %s", currentGK.Group, desired.GetName()),
+		),
+		)
+	}
+	if currentGK.Kind != desiredGK.Kind {
+		allErrs = append(allErrs, field.Invalid(
+			field.NewPath("metadata", "kind"),
+			desiredGK.Kind,
+			fmt.Sprintf("must be equal to %s for %s", currentGK.Kind, desired.GetName()),
+		),
+		)
 	}
-	return ObjectsAreInTheSameNamespace(current, desired)
+	allErrs = append(allErrs, ReferencedObjectsAreInTheSameNamespace(current, desired)...)
+	return allErrs
 }
 
-// ObjectsAreInTheSameNamespace checks if two referenced objects are in the same namespace.
-func ObjectsAreInTheSameNamespace(current, desired client.Object) error {
+// ReferencedObjectsAreInTheSameNamespace checks if two referenced objects are in the same namespace.
+func ReferencedObjectsAreInTheSameNamespace(current, desired client.Object) field.ErrorList {
+	var allErrs field.ErrorList
+
 	// NOTE: this should never happen (webhooks prevent it), but checking for extra safety.
 	if current.GetNamespace() != desired.GetNamespace() {
-		return errors.Errorf("invalid operation: it is not possible to change the namespace of %s/%s from %s to %s",
-			current.GetObjectKind().GroupVersionKind(), current.GetName(), current.GetNamespace(), desired.GetNamespace())
+		allErrs = append(allErrs, field.Invalid(
+			field.NewPath("namespace"),
+			desired.GetNamespace(),
+			fmt.Sprintf("must be equal to %s for %s", current.GetNamespace(), desired.GetName()),
+		))
+	}
+	return allErrs
+}
+
+// ReferencedTemplatesAreCompatible checks if two referenced objects are compatible, meaning that
+// they are of the same GroupKind and in the same namespace.
+func ReferencedTemplatesAreCompatible(current, desired clusterv1.LocalObjectTemplate) field.ErrorList {
```

Comment:
Should this be 
```suggestion
func TemplateReferencesAreCompatible(current, desired clusterv1.LocalObjectTemplate) field.ErrorList {
```

Reply:
Should we maybe change this to References are compatible and just check the references? That would maybe clean up some of the semantic issues across these functions.

Reply:
You mean only pass in the refs? Seems okay to me

Reply:
Let's keep LocalObjectTemplate, so we are ready if we add more field to this struct.
WRT to the name, it is a nit.
- I'm -1 with `ReferencedTemplatesAreCompatible` because it seems that we are looking at the actual template
- `TemplateReferencesAreCompatible` kind of solve the problem IMO
- might be `LocalObjectTemplateAreCompatible` addresses it definitely ;-)